### DISCUSSION TITLE
Config File Use Case (for AT Projects, but generally useful)

### DISCUSSION THREAD

**Discussion by ATMakersBill** [adafruit/Adafruit_CPFS#3] [General]
Thanks for starting this repo @PaintYourDragon 

Just to get it down on paper, in the ATMakers community, we have a regular problem where we need to let the end-user configure their devices.  On CircuitPython, this is simple because we can just have a settings.py file that has constants that are included in main.py and let the user edit settings.py.  There are downsides to this on memory constrained boards like the SAMD21, but in general it works well.  It has also led us to use CircuitPython on projects where Arduino (or the C/C++ SDK for the RP2040) may have been a better choice in terms of speed, etc.  Our Sip&Puff Morse Code project was one such example, and our MouseFilter projects (IntelliKeys, TremorMouse, etc.) are all similar situations - even if the prototyping can be done in CircuitPython, there are times when the speed of compiled code will just help the overall experience.

We have tried other config options.  Chris Young (@cyborg5) tried to use the Arcada project's config file system and had issues here https://forums.adafruit.com/viewtopic.php?f=25&t=176032 and tried to work around them.  But it wasn't reliable in the long run.  We've also tried WebUSB, WebSerial, and WebBT... all were problematic.

With this library, we would probably use one of two methods to manage the config file: JSON files or INI files.  JSON files would probably use the ArduinoJSON library ( https://arduinojson.org/ ) and INI files would use MiniIni4Arduino ( https://github.com/lipoyang/minIni4Arduino ).  I believe in most cases, we would use the INI files - they seem easier for end-users to read and edit and we don't need often need the structure JSON can provide.

Some ideal features that we would use would be:

- Having a call-back that was triggered when the INI file was modified.  Ideally this would be on file change, not last-viewed, etc. but we can always filter things.  This would re-read the file when it was changed by the user.
- Being able to initially create the file with default values if it was missing.  That means write-access from the Arduino code at least in certain cases.
- Being able to enable and disable the MSD endpoint.  We would like (in some cases) to have a slide-switch or button that triggers "setup mode" which makes the file system show up on the host computer while the default case is that the file system is not exposed.

Hope this helps

**Comment by PaintYourDragon** [comment]
Regarding the last item (button/switch to enable filesystem or not) — if it’s sufficient to _only_ do this on startup/reset, then it looks like a pretty simple matter, I could add an argument to begin(). If this needs to work “live,” I’m not certain if TinyUSB supports that or if USB mounting is strictly a one-time operation. There’s a few functions that look like possibilities but I’ve never poked them with a stick.

**Reply by ATMakersBill** [reply]
I think holding a button while starting would be fine.  As for whether it's possible, it looks like @brentru uses .detach() and .attach() to accomplish this, but I'm not certain.

**Comment by PaintYourDragon** [comment]
I’ll take a look at attach/detach as an option…but at the very least, the button-on-startup should be simple enough.

As for settings files: I think that’s beyond the scope of this. While the main use likely _is_ settings files, really want to remain agnostic as to _what_ settings format(s). Already have upcoming projects involving XML and JSON, and lately CircuitPython is adopting TOML for certain global configurables, and there’s still others. Other projects, it’s not about settings at all, but media files. Think it’s best to stand back and let the application do what it needs to do in that regard. The library’s just here because I hated putting all that flash-mounting code in every sketch.

**Reply by ATMakersBill** [reply]
Yes: the format and parsing and stuff would be on top of this. I put this in discussions just so you’d know what we would use it for and keep that in mind.
